drv/lpc55-sprot-server/src/main.rs

```diff
@@ -149,63 +153,55 @@ enum IoError {
-        let mut rx_msg = RxMsg::new(&mut rx_buf[..]);
-
-        let reply = match io.next(signal_reply, &tx_buf[..], &mut rx_msg) {
-            Ok(()) => {
-                // Put a busy byte in the tx_fifo before we handle the result,
-                // which may take some time.
-                io.mark_busy();
-                // Clear the buffer in preparation for a reply
-                let tx_msg = TxMsg::new(&mut tx_buf[..]);
-                handler.handle(rx_msg, tx_msg, &mut io.stats)
+        let rsp_len = match io.wait_for_request(&mut rx_buf) {
+            Ok(rx_len) => {
+                handler.handle(&rx_buf[..rx_len], &mut tx_buf, &mut io.stats)
+            }
+            Err(IoError::Flush) => {
+                // A flush indicates that the server should de-assert ROT_IRQ
+                // as instructed by the SP. We do that and then proceed to wait
+                // for the next request.
+                ringbuf_entry!(Trace::Flush);
+                io.deassert_rot_irq();
+                continue;
             }
-            Err(IoError::Flush) => None,
             Err(IoError::Flow) => {
-                io.mark_busy();
-                let tx_msg = TxMsg::new(&mut tx_buf[..]);
-                Some(handler.flow_error(tx_msg))
+                ringbuf_entry!(Trace::FlowError);
+                handler.flow_error(&mut tx_buf)
             }
         };
 
-        if reply.is_none() {
-            // There's no reply to send, so we should clear the Tx buffer.
-            tx_buf.fill(0);
-            signal_reply = false;
-        } else {
-            signal_reply = true;
-        }
+        ringbuf_entry!(Trace::Stats(io.stats));
+        io.reply(&tx_buf[..rsp_len]);
     }
 }
 
 impl Io {
-    /// Put a Protocol::Busy value in FIFOWR so that SP/logic analyzer knows
-    /// we're away.
-    pub fn mark_busy(&mut self) {
-        self.spi.drain_tx();
-        self.spi.send_u8(Protocol::Busy as u8);
-    }
-
-    pub fn next<'a>(
-        &mut self,
-        signal_reply: bool,
-        tx_buf: &[u8],
-        rx_msg: &mut RxMsg<'a>,
-    ) -> Result<(), IoError> {
-        let mut tx_iter = tx_buf.iter();
-        self.prime_tx_fifo(&mut tx_iter);
-        let result = loop {
+    // Wait for chip select to be asserted
+    // Assert ROT_IRQ if this is a reply
+    fn wait_for_csn_asserted(&mut self, is_reply: bool) {
+        // This is just an optimization in order to prevent having to call the
+        // gpio task when its not necessary. If the server crashes, and ROT_IRQ
+        // is already asserted, and we somehow get here, we'll just make one
+        // more idempotent call to the gpio server.
+        let mut rot_irq_asserted = false;
+        loop {
             sys_irq_control(notifications::SPI_IRQ_MASK, true);
 
-            if signal_reply {
+            if is_reply && !rot_irq_asserted {
+                rot_irq_asserted = true;
```

Comment:
Should `rot_irq_asserted` be a member variable, instead of just being local to this function?

Reply:
I like this better. Added in [3ddb01b](https://github.com/oxidecomputer/hubris/pull/1279/commits/3ddb01b536d37432e9c62a391d4b708a2630ce23)